### PR TITLE
Make the Stage insertion API in CodePipeline more flexible

### LINKED ISSUE #347: codepipeline: There should be a way to insert a stage before an existing one (and not append)
Currently, Pipelines only allows appending stages. This creates a limitation in usability. For example, we have a general purpose construct that defines a pipeline for code library builds. The current pipeline has two stages: build and release. We wish to implement a method `.addTest(...)`, which will conditionally "inject" another stage between "build" and "release" only if it's called. If there are no tests, there should still be two stages.

Maybe we can do something like `new Stage(pipeline, 'Bla', { insertBefore: someOtherStage })`.

### PR DESCRIPTION
**Pull request by skinny85**
Allows inserting Stages into CodePipeline at an arbitrary index, and also before or after any given Stage.

--------------------------------------------------------

Fixes #347 

--------------------------------------------------------

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) license.

### CODE REVIEW

**Comment by skinny85** [comment]
@eladb ping (you've opened the original issue, after all ;) )

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 1]
```diff
@@ -131,6 +131,20 @@ export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
...
+     * Get a duplicate of this Pipeline's list of Stages.
+     */
+    public get stages(): Stage[] {
+        return this._stages.slice();
+    }
+
+    /**
+     * Get the number of Stages in this Pipeline.
+     */
+    public get stagesLength(): number {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: stageCount

**Comment by eladb** [comment]
@skinny85 sorry! I provided some feedback the previous PR but forgot to "submit feedback". Copying the notes to this PR now. Apologies...

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 2]
```diff
@@ -216,23 +230,71 @@ export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
      *       onChildAdded type hook.
      * @override
      */
-    protected addChild(child: cdk.Construct, name: string) {
-        super.addChild(child, name);
+    protected addChild(child: cdk.Construct, name: string, props?: any) {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: I don’t think this change is needed and we are extremely sensitive to extensions of the surface area of Construct. 

I see why you need to propagate props to `addChild` but that’s not a robust way to do this. What promises you that everyone will pass in their props to `Construct`?

I think the way to address this is to render the stages array only during synthesis (using a `Token`). Then, you wouldn’t even need to override `addChild` at all. When you synthesize, you will iterate over your children and devise the list of stages according to their placement.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 3]
```diff
@@ -216,23 +230,71 @@ export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
...
 
-        this.stages.push(stage);
+        const index = placement
+            ? this.calculateInsertIndexFromPlacement(placement)
+            : this.stagesLength;
+
+        this._stages.splice(index, 0, stage);
+    }
+
+    private calculateInsertIndexFromPlacement(placement: StagePlacement): number {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Can we validate `stagePlacement` in the ctor so it's ensured to be correct when we get here?

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts [hunk 4]
```diff
@@ -5,6 +5,52 @@ import { cloudformation } from './codepipeline.generated';
...
+/**
+ * Allows you to control where to place a new Stage when it's added to the Pipeline.
+ * Note that you can provide only one of the below properties -
+ * specifying more than one will result in a validation error.
+ *
+ * @see #rightBeforeStage
+ * @see #justAfterStage
+ * @see #atIndex
+ */
+export interface StagePlacement {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Not sure I think the parent/child semantics make total sense here. Why not just “before” and “after”? What am I missing?

File: packages/@aws-cdk/cdk/lib/core/construct.ts [hunk 5]
```diff
@@ -35,9 +35,10 @@ export class Construct {
      * Creates a new construct node.
      *
      * @param parent The parent construct
-     * @param props  Properties for this construct
+     * @param name The logical identfifier of the new Construct
+     * @param props Optional properties for this construct
      */
-    constructor(parent: Construct, name: string) {
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/core/construct.ts]: Please revert this change to `Construct`

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 1 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: 👍🏻

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 2 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Changes in Construct are not needed anymore after #568.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Can you elaborate how is that better? IMO, splitting validation and usage that assumes that validation has been done makes for less readable code.

File: packages/@aws-cdk/cdk/lib/core/construct.ts (hunk 5 above)
Comment (skinny85) [reply packages/@aws-cdk/cdk/lib/core/construct.ts]: As said above, not needed anymore.

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts (hunk 4 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: It's just a matter of wording. I wanted to make it clear that inserting a Stage before/after a given Stage will "displace" any stage that was before/after that Stage there before (see how clunky this explanation is?), and that's why I came up with the child/parent wording. I'm open to suggestions on how to improve those docs, of course.

**Comment by skinny85** [comment]
Well then. After #568, this change got a whole lot easier.

**Comment by skinny85** [comment]
@eladb ping :)

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 6]
```diff
@@ -210,6 +210,20 @@ export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
         ]);
     }
 
+    /**
+     * Get a duplicate of this Pipeline's list of Stages.
+     */
+    public get stages(): Stage[] {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Is this really needed? (Keep surface area to minimum)

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts [hunk 7]
```diff
@@ -5,6 +5,39 @@ import cdk = require('@aws-cdk/cdk');
...
+ * @see #rightBefore
+ * @see #justAfter
+ * @see #atIndex
+ */
+export interface StagePlacement {
+    /**
+     * Inserts the new Stage as a parent of the given Stage
+     * (changing its current parent Stage, if it had one).
+     */
+    readonly rightBefore?: Stage;
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Any reason “before” and “after” are not sufficient (keep names concise)

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: It’s a principle in contract-driven design: Ideally you want to validate as close as possible to the surface area and then the rest of the code can assume invariants are kept. It has a few benefits:

1. You only validate once: if in the future someone would add some other code path will want to use this value, they won’t have to duplicate the validation
2. Users see the error “closer” to where they made the mistake.

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts [hunk 8]
```diff
@@ -13,6 +46,14 @@ export interface StageProps {
...
      */
     pipeline: Pipeline;
+
+    /**
+     * Allows specifying where should the newly created {@link Stage}
+     * be placed in the Pipeline.
+     *
+     * @default the stage is added at the end of the Pipeline
+     */
+    placed?: StagePlacement;
```
Comment (eladb) [inline packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: I would call this placement. Our properties are usually nouns and not verbs

Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: I named it so that the code reads a little better (`new Stage(this, 'Stage', { placed: { atIndex: 3 }})`), but I can change it to `placement` if you prefer that.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 3 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: > 2. Users see the error “closer” to where they made the mistake.

That's already the case here - `calculateInsertIndexFromPlacement`is called from `_addStage`, which in turn is called in the `Stage` constructor. So the user will see the error immediately when constructing the invalid Stage (we have a unit test confirming this behavior).

Since I'm fairly sure this parameter will not be used anywhere else, I'll leave this as it is right now, if it's OK with you.

File: packages/@aws-cdk/aws-codepipeline/lib/stage.ts (hunk 8 above)
Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: As I said, properties are usually nouns and not verbs. Just trying to keep our semantic style uniform

Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/stage.ts]: Will fix.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts (hunk 6 above)
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Sure. Will remove.

**Comment by skinny85** [comment]
> Any updates on this?

Incorporated feedback from your last review, and rebased.

File: packages/@aws-cdk/aws-codepipeline/README.md [hunk 9]
```diff
@@ -16,6 +16,21 @@ const sourceStage = new Stage(this, 'Source', {
 });
 ```
 
+You can insert the new Stage at an arbitrary point in the Pipeline:
+
+```ts
+const sourceStage = new Stage(this, 'Source', {
+    pipeline,
+    placement: {
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/README.md]: Yes, but we have to merge either this, or the `addStage` PR first (they're both based on `master`). Once I do that, I'll rebase the other one on top of the new `master`, and edit accordingly.

File: packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts [hunk 10]
```diff
@@ -208,8 +215,9 @@ export class Pipeline extends cdk.Construct implements events.IEventRuleTarget {
      * so there is never a need to call this method explicitly.
      *
      * @param stage the newly created Stage to add to this Pipeline
+     * @param placement an optional specification of where to place the newly added Stage in the Pipeline
      */
-    public _addStage(stage: Stage): void {
+    public _addStage(stage: Stage, placement?: StagePlacement): void {
```
[the remark this answers is not in the document]
Comment (skinny85) [reply packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts]: Will do.

File: packages/@aws-cdk/aws-codepipeline/README.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/aws-codepipeline/README.md]: Cool. Both are approved.

**Comment by skinny85** [comment]
Rebased on top of the merged `Pipeline#addStage` PR, incorporating the latest feedback.
